app/javascript/controllers/select_with_auto_complete_controller.js

```diff
@@ -0,0 +1,428 @@
+    var categoryItems = category.querySelectorAll(':scope > [role="option"]');
+    for (var i = 0; i < categoryItems.length; i++) {
+      var categoryItem = categoryItems[i];
+      if (add) {
+        this.allOptions.push(categoryItem);
+      }
+      this.addListboxOptionEventListeners(categoryItem);
+    }
+  }
+
+  getLowercaseContent(node) {
+    return node.textContent.toLowerCase();
+  }
+
+  isOptionInView(option) {
+    var bounding = option.getBoundingClientRect();
+    return (
+      bounding.top >= 0 &&
+      bounding.left >= 0 &&
+      bounding.bottom <=
+        (window.innerHeight || document.documentElement.clientHeight) &&
+      bounding.right <=
+        (window.innerWidth || document.documentElement.clientWidth)
+    );
+  }
+
+  setActiveDescendant(option) {
+    if (option) {
+      this.comboboxTarget.setAttribute("aria-activedescendant", option.id);
+      if (!this.isOptionInView(option)) {
+        option.scrollIntoView({ behavior: "smooth", block: "nearest" });
+      }
+    } else {
+      this.comboboxTarget.removeAttribute("aria-activedescendant");
+    }
+  }
+
+  setValue(option) {
+    this.hiddenTarget.value = option ? option.getAttribute("data-value") : "";
+    this.filter = option ? option.textContent : "";
+    this.comboboxTarget.value = this.filter;
+    this.comboboxTarget.setSelectionRange(
+      this.filter.length,
+      this.filter.length,
+    );
+    this.filterOptions();
+    this.setOption(null);
+  }
+
+  // ComboboxAutocomplete events
+
+  filterOptions() {
+    var option = null;
+    var currentOption = this.option;
+    var filter = this.filter.toLowerCase();
+
+    this.filteredOptions = [];
+    this.listboxTarget.innerHTML = "";
+
+    for (var i = 0; i < this.allOptions.length; i++) {
+      var flag = false;
+      var optionCategory = this.allOptions[i].cloneNode(true);
+
+      if (optionCategory.role === "group") {
+        var options = optionCategory.querySelectorAll('[role="option"]');
+        for (var j = 0; j < options.length; j++) {
+          option = options[j];
+          this.addListboxOptionEventListeners(option);
+          if (
+            filter.length === 0 ||
+            this.getLowercaseContent(option).indexOf(filter) >= 0
+          ) {
+            flag = true;
+            var regex = new RegExp(`(${this.filter})`, "gi");
+            option.innerHTML = option.textContent.replace(
+              regex,
+              "<mark class='bg-primary-300 dark:bg-primary-600 font-semibold'>$1</mark>",
+            );
+            this.filteredOptions.push(option);
+          } else {
+            optionCategory.removeChild(option);
+          }
+        }
+      } else {
+        this.addListboxOptionEventListeners(optionCategory);
+        if (
+          filter.length === 0 ||
+          this.getLowercaseContent(optionCategory).indexOf(filter) >= 0
+        ) {
+          flag = true;
+          var regex = new RegExp(`(${this.filter})`, "gi");
```

Comment:
The `this.filter` value (user input) is directly interpolated into a RegExp without escaping special regex characters. Malicious input like .* could cause ReDoS (Regular Expression Denial of Service) or unexpected matching behavior.

Make a helper method:
```javascript
escapeRegExp(string) {
    return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
 }
```

Then:
```javascript
var regex = new RegExp(`(${this.escapeRegExp(this.filter)})`, "gi");
```